### ISSUE TITLE
Saved image colors don't honor color table

### ISSUE THREAD

**Issue by markcmiller86** [visit-dav/live-customer-response#76]
We're trying to [combine `visit-users` email list with GitHub issues](https://github.com/visit-dav/live-customer-response/wiki/How-the-new-GitHub-visit-users-Email-Integration-Works).
When replying on `visit-users`, please reply **only** to emails with a *Subject*
line that includes `[visit-dav/live-customer-response]`.

---

I'm using VisIt 3.0.0 and I'm trying to plot magnetic field lines with the integral curve operator on a pseudocolor plot. However, sometimes the lines aren't colored correctly, meaning they don't match with the colortable set in the pseudocolor plot. I am only using the cli and python scripting to do this, no gui whatsoever. The odd thing is, I printed the pseudocolor attributes periodically throughout the script, including just after DrawPlots and just after SaveWindow and they all say that the colortable is set correctly. But the saved image has the lines as red (instead of white). I tried using both a custom colortable and a premade colortable (gray) but they still came out as red (I'm assuming this is the default colortable, hot).

In the same image, we also load in density (matter) data and create a second pseudocolor plot with a different custom colortable for this density data. Here, the colortable is set properly and the image looks fine, so the issue seems to only be happening when I add the integralcurve operator. Also note that we didn't have this problem when using VisIt 2.9.2 and the Streamlines plot. So I'm really at a loss for what is causing this issue. Is there a bug in 3.0.0 that I don't know about that could be causing this? I checked the documentation and it didn't seem like it.

**Comment by markcmiller86** [comment]
Kyle, first, would it easy for you to upgrade to [VisIt 3.1.2](https://github.com/visit-dav/visit/releases/tag/v3.1.2)?

Also, what `Color by` option are you using in the IC operator?

**Comment by knelli2** [comment]
Unfortunately not. I am working on NASA's supercomputer Pleiades, and the only versions of VisIt they have installed are 2.13.0 and 3.0.0. We are currently trying to have support as NASA install other versions but I don't know if that will happen.

I don't seem to be able to find the `Color by` attribute in the IC Attributes. For the Pseudocolor plot though, we are using our own colortable (attached) and have have the opactiy as FullyOpaque.

Attached are the two main scripts we use: `run.py` and `runModule.py`. `run.py` is the actual script that gets run by VisIt, but `runModule.py` has most of the actual plotting features. The main function in `runModule.py` where we setup the Pseudocolor and IntegralCurve is called `PlotB` given here:

```
def PlotB(database, indx, ref=1):
        ActivateDatabase(database)
        AddPlot("Pseudocolor","BVec")   #plot 1
        print("Add streamline plot with index = {}".format(indx))`

        SetActivePlots(indx)
        if ref:
                reflect()

        pseudoAtt = PseudocolorAttributes()
        pseudoAtt.legendFlag = 0
        pseudoAtt.lightingFlag = 1
        pseudoAtt.colorTableName = 'streamlines'
        #pseudoAtt.lineType= pseudoAtt.Ribbon
        pseudoAtt.lineWidth = 3
        SetPlotOptions(pseudoAtt)
        print(pseudoAtt)
        AddOperator("IntegralCurve")
        return IntegralCurveAttributes()
```
Later in the script we load in an IntegralCurve.xml file into the attributes (also attached).

Also, I was curious about the `lineType` attribute of pseudocolor because I had always been using `Line` so I tried `Tube` and `Ribbon` and actually using ribbons gave me the correct colored plot, but lines and tubes still gave me red plots (attached images are the exact same except for the `lineType` attribute).

[all_attachments.zip](https://github.com/visit-dav/live-customer-response/files/4733313/all_attachments.zip)

**Comment by markcmiller86** [comment]
@knelli2 apologies for delays in responding. Will investigate later today. Thanks for attaching script to test.

**Comment by markcmiller86** [comment]
@knelli2 ok just debugging this now. Haven't been able to run your script just yet but am getting closer. A few questions...

From this command line...

```
#visit -cli -nowin -forceversion 2.7.3 -s $visitScript $dir $xmldir $saveFolder $rank \
$totranks $streamXML $vecXML $bsqXML $maxdensity $rho_pseudoXML $rho_isoXML $g00_pseudoXML $g00_isoXML
```

`$dir` is arg index `1` once in VisIt's CLI Python interpreter. `$xmldir` is arg index `2` and so forth. However, the hard-coded array indices used in `run.py` seem to be using `13` and `14` to access those arg indices and this is confusing a me a tad. Can you explain?

When you are running this client/server, do you have the custom color table installed in both the client and server or only on client? I dunno if this will effect things...I am testing something similar myself to see if it is an effect.

**Comment by markcmiller86** [comment]
> When you are running this client/server, do you have the custom color table installed in both the client and server or only on client? I dunno if this will effect things...I am testing something similar myself to see if it is an effect.

Answering my own question here...VisIt will get this right. You do NOT have to copy the color table file to the server machine.

**Comment by knelli2** [comment]
@markcmiller86 Sorry about the confusion. We've fallen victim to not keeping the comments in our code fully up to date. We've updated the arguments submitted to our python scrip since that comment was written. We can a somewhat complicated setup built to submit jobs but the main command we use to run visit is this:

`visit -cli -nowin -forceversion 3.0.0 -s $visitScript $PlotDensAsVol $PlotDensAsIso $PlotDensLinear $PlotVel $PlotBsq2r $Plotg00 $refPlot $cutPlot $PlotEvolve $PlotZoom $PlotFlyOver $PlotFlyAround $dir $xmldir $tosave$(printf "%03d" $rank)"_" $rank $totranks $numBfieldPlots $vecXML $bsqXML $maxdensity $rho_pseudoXML $rho_isoXML $g00_pseudoXML $g00_isoXML`

This should be more in line with what is found in `run.py`.

As for you other question, we store our custom colortable in the `~/.visit/` folder in whatever environment we are in (local computers or supercomputer).

**Comment by markcmiller86** [comment]
Ok, I've tried to duplicate the behvior you are seeing by running 3.0.1 client on OSX and launching a 3.0.2 server (serial engine) on a remote resource in Livermore Computing. The script I am running is very simple, using a test data file, `globe.silo`. Here is my script...

```
import sys
OpenDatabase("surface.llnl.gov:/usr/gapps/visit/data/globe.silo")

# set background to grey so we can see white streamlines
annotAtts = AnnotationAttributes()
annotAtts.backgroundColor = (128, 128, 128, 255)
SetAnnotationAttributes(annotAtts)

# Do a PC plot of an integral curve variable (this is different from your case)
AddPlot("Pseudocolor","operators/IntegralCurve/vel")

# Set plot color table
pcAtts = PseudocolorAttributes()
pcAtts.colorTableName = 'streamlines'
SetPlotOptions(pcAtts)

# Adjust IC attrs a bit
icAtts = IntegralCurveAttributes()
icAtts.sourceType = icAtts.SpecifiedSphere
icAtts.radius = 10
icAtts.randomSamples = 1
icAtts.numberOfRandomSamples = 100
SetOperatorOptions(icAtts)

# Lets see what we get
DrawPlots()
SaveWindow()
sys.exit()
```

I get an image with white streamlines. So, maybe the issue is in how you are instantiating the intergral curve operator. Are you doing so by going through the Pseudocolor plot's `operator->IntegralCurve` sub-menu or are you building it yourself? I think the latter approach is frought with peril due to what is involved in setting the IC Operator up.

![image](https://user-images.githubusercontent.com/5720676/84466159-9d40d900-ac2d-11ea-952e-3f4429cf677e.png)

**Comment by knelli2** [comment]
```
> def PlotB(database, indx, ref=1):
>         ActivateDatabase(database)
>         AddPlot("Pseudocolor","BVec")   #plot 1
>         print("Add streamline plot with index = {}".format(indx))
> 
>         SetActivePlots(indx)
>         if ref:
>                 reflect()
> 
>         pseudoAtt = PseudocolorAttributes()
>         pseudoAtt.legendFlag = 0
>         pseudoAtt.lightingFlag = 1
>         pseudoAtt.colorTableName = 'streamlines'
>         #pseudoAtt.lineType= pseudoAtt.Ribbon
>         pseudoAtt.lineWidth = 3
>         SetPlotOptions(pseudoAtt)
>         print(pseudoAtt)
>         AddOperator("IntegralCurve")
>         return IntegralCurveAttributes()
```
From one of my previous replies, this is the function we use. `BVec` is a simple vector expression created from three scalar fields `Bx`, `By`, `Bz`. We create a PC plot of `BVec`, set all the PC attributes, add an IC operator, and then return an ICAttributes object to be used later in the script to load in a custom pointlist. (This can all be found in `runModule.py` if you want to take a look`)

Are you suggesting, instead of adding a PC plot then adding an operator, adding a PC plot alread instantiated with an IC operator? Namely

`AddPlot("Pseudocolor","operators/IntegralCurve/BVec")`

instead of

```
AddPlot("Pseudocolor","BVec")
AddOperator("IntegralCurve")
```

I should note that we also downloaded VisIt 3.0.2 on our local machines and tried these scripts and they loaded in the custom colortable perfectly fine, regardless of what `lineType` I chose.

**Comment by markcmiller86** [comment]
Ok, lemme add the use of an *expression* to the process to see if that might make a difference. Then, I'll try explicitly manipulating the IC operator as you have too.

**Comment by markcmiller86** [comment]
Ok, I've tried various combinations of *expressions* as well as *explicitly* adding the IC operator and in all my tests, the coloring of the streamlines still comes out white.

Can you add some calls to get and print active plot options just before `DrawPlots()`

```
pcAtts = GetPlotOptions()
print(pcAtts.colorTableName)
DrawPlots()
```

**Comment by knelli2** [comment]
[bringleout.txt](https://github.com/visit-dav/live-customer-response/files/4772365/bringleout.txt)
I actually did that as one of my tests before I came here. This file contains the output from our script when it is run. I printed out the PC attributes at multiple stages of the script. You'll want to search for the phrases

`after operator options is set`
`after DrawPlots`
`after SaveWindow`

These are the places when I printed all the PC attributes. `after operator options is set` is basically right before `DrawPlots()`, then I print it `after DrawPlots`, and finally `after SaveWindow`. Every time, the the colorTableName is correct. (Note: In this run, I copied streamlines.ct to streamlines2.ct to test if my colortable was corrupted somehow. So you'll see `streamlines2` as the colortable name instead of just `streamlines`, but they are identical.) (Note2: This only happens when `pcAtts.lineType = pcAtts.Line` or `pcAtts.lineType = pcAtts.Tube`. If `pcAtts.lineType = pcAtts.Ribbon`, the integral curves are white, like they are supposed to be.)

**Comment by markcmiller86** [comment]
Ok, so here is what I have most recently tested, including setting `pcAtts.lineType = pcAtts.Line`. Pls copy this code to `foo.py` and adjust `OpenDatabase(...` line to specify correct machine name and file path for `globe.silo`. Am attaching the test data file too, `globe.silo` so you can perform same test.

```
import sys
DefineVectorExpression("myvec","{dx, dx, dx}")
OpenDatabase("surface.llnl.gov:/usr/gapps/visit/data/globe.silo")

annotAtts = AnnotationAttributes()
annotAtts.backgroundColor = (128, 128, 128, 255)
SetAnnotationAttributes(annotAtts)

#AddPlot("Pseudocolor","operators/IntegralCurve/vel")
AddPlot("Pseudocolor","myvec")
pcAtts = PseudocolorAttributes()
pcAtts.colorTableName = 'streamlines'
pcAtts.lineType = pcAtts.Line
SetPlotOptions(pcAtts)

AddOperator("IntegralCurve")
icAtts = IntegralCurveAttributes()
icAtts.sourceType = icAtts.SpecifiedSphere
icAtts.radius = 10
icAtts.randomSamples = 1
icAtts.numberOfRandomSamples = 100
SetOperatorOptions(icAtts)

DrawPlots()
SaveWindow()
sys.exit()
```

Then run from your local machine like so...

```
/Applications/VisIt.app/Contents/Resources/bin/visit -cli -nowin -debug 5 -s ./foo.py
```

[globe.silo.gz](https://github.com/visit-dav/live-customer-response/files/4772450/globe.silo.gz)

**Comment by knelli2** [comment]
So I ran the sample script you gave me with globe.silo and it came out correct. The lines look a tad weird, but they are white.
![mark_test_line0000](https://user-images.githubusercontent.com/31455476/84540217-80061c00-acba-11ea-9b89-5d8e5bb2b3a3.png)

In the debug files `A.viewer.4.vlog` and `A.viewer.5.vlog` there is the line
```
Imported color table /home4/knelli/.visit/streamlines.ct as streamlines
```
so it seems to be importing the colortable correctly. I tried running our code with -debug 5 as a flag, both with ribbons and lines, and the above line appears in both runs, in both `A.viewer.4.vlog` and `A.viewer.5.vlog` exactly as it did for your test script, yet the lines still turned red, but the ribbons were white. However, I then tried adding the IC operator when I added the PC plot and the fields came out white! But they weren't as long

Lines:
```
AddPlot("Pseudocolor","BVec")
AddOperator("IntegralCurve")
```
![high_test13_debugging_lines_001_000_0000](https://user-images.githubusercontent.com/31455476/84540482-10dcf780-acbb-11ea-8da3-501e97488412.png)

`AddPlot("Pseudocolor","operators/IntegralCurve/BVec")`
![high_test13_debugging_lines_instantiate_operator_001_000_0000](https://user-images.githubusercontent.com/31455476/84548225-c8c5d100-acca-11ea-9072-58713ab95e5d.png)


Ribbons:
```
AddPlot("Pseudocolor","BVec")
AddOperator("IntegralCurve")
```
![high_test13_debugging_ribbons_001_000_0000](https://user-images.githubusercontent.com/31455476/84540488-15a1ab80-acbb-11ea-8b67-4b2ecb039a6b.png)

`AddPlot("Pseudocolor","operators/IntegralCurve/BVec")`
![high_test13_debugging_ribbons_instantiate_operator_001_000_0000](https://user-images.githubusercontent.com/31455476/84548268-dd09ce00-acca-11ea-93a2-c1c5caab12c5.png)


Everything else was exactly the same. Apparently the order matters in my case for some reason I don't understand. And I don't know why the lines terminated prematurely when I added the IC operator with the PC plot.

**Comment by markcmiller86** [comment]
Ok, those white lines from `globe.silo` look correct. Yeah, they are funny. Its a contrived vector expression which has the same variable for all three components.

Thanks for running that test. That explains why I am unable to reproduce also.

@ARSanderson can you maybe lend an eye here and suggest anything that might be going amiss with IntegralCurve operator under these conditions?

**Comment by markcmiller86** [comment]
Ok, I'v modified my test further trying to narrow in on a reproducer (for color issue or shorter line issue you're now seeing). I've adjusted my IC operator to use a point list and I've adjusted the annotation attributes a bit more. Am also adding a volume plot, which I think you have there aslo for the center object.

Forgot to ask, where is the red coloring coming from for the torus shaped object at the center? I mean, is that the correct coloring there?

**Comment by knelli2** [comment]
The red object in the middle is a star which can be approximated to an oblate spheroid. We use a volume plot with RayCasting to represent this. The volume plot has a custom xml with color control points and custom opacity profile. Yes it is the correct color.

**Comment by markcmiller86** [comment]
Ok, my latest attempt to reproduce is now using also a volume plot

```
import sys
#OpenMDServer("surface.llnl.gov")
#OpenDatabase("/usr/gapps/visit/data/globe.silo")
#OpenDatabase("/Users/miller86/visit/visit/data/silo_hdf5_test_data/globe.silo")
DefineVectorExpression("myvec","{dx, dx, dx}")
OpenDatabase("surface.llnl.gov:/usr/gapps/visit/data/globe.silo")

annotAtts = AnnotationAttributes()
annotAtts.backgroundColor = (128, 128, 128, 255)
SetAnnotationAttributes(annotAtts)

AddPlot("Volume", "speed")
volAtts = VolumeAttributes()
LoadAttribute("volume_plot.xml", volAtts)
SetPlotOptions(volAtts)

#AddPlot("Pseudocolor","operators/IntegralCurve/vel")
AddPlot("Pseudocolor","myvec")
pcAtts = PseudocolorAttributes()
pcAtts.colorTableName = 'streamlines'
pcAtts.lineType = pcAtts.Line
SetPlotOptions(pcAtts)

AddOperator("IntegralCurve")
icAtts = IntegralCurveAttributes()
icAtts.sourceType = icAtts.PointList
icAtts.pointList = (1,1,1,3,3,3,5,5,5,7,7,7,0,0,1,1,1,1,2,2,1,3,3,1)
SetOperatorOptions(icAtts)

DrawPlots()
SaveWindow()
sys.exit()
```

which still gives white for the one and only streamline it winds up producing 😕 

![image](https://user-images.githubusercontent.com/5720676/84553989-c4011d00-acca-11ea-979e-775c5e8cecc9.png)

**Comment by knelli2** [comment]
Haha I love when VisIt decides to plot only one of the many streamlines you had.

Well plotting the PC plot and IC operator all at once fixed the color issue, but now I'm not sure why they are terminating prematurely when they were integrating to full length before (regardless of line/tube/ribbon). I tried doubling the `maxSteps` in the ICAttributes, but that didn't do anything so now I'm at a loss again.

**Comment by markcmiller86** [comment]
> Haha I love when VisIt decides to plot only one of the many streamlines you had.

I suspect that is an artifact of the particular points I specified and the contrived vector field I was working with. I specified points almost *randomly* w.r.t. actual dataset geometry. As long as I was generating a streamline of some kind to reveal coloring, I was fine. That doesn't help debug possible early termination though.

**Comment by ARSanderson** [comment]
Hello, I have nothing substantive to add. Other than when using Python adding the plot with

AddPlot("Pseudocolor","BVec")
AddOperator("IntegralCurve")

or

AddPlot("Pseudocolor","operators/IntegralCurve/BVec")

should in theory both work. The latter is needed simply for the GUI which restricts showing variables that are only possible for that plot type. 

However, I could see for the former that PC plot during some initialization step horks because the data it initially has is vector data and has no clue what to do. But then once the IC operator is added which produces scalar that state remains horked. As such, I would change the ordering to adding the operator BEFORE mucking with PC attributes.

AddPlot("Pseudocolor","BVec")
AddOperator("IntegralCurve")
... Set plot color table
pcAtts = PseudocolorAttributes()
pcAtts.colorTableName = 'streamlines'
SetPlotOptions(pcAtts)

**Comment by markcmiller86** [comment]
I've tried several other things...

- Added Camera light and set specific direction
- Tried various of the ray-casting volume plot methods for the volume plot
- Changed order of Volume plot relative to PC plot
- Tried both ribbon and line type for PC plot

I've also switched to `vel` variable in `globe.silo` (which is not an expression) which is a more *normal* vector field. After updating the point list, I get streamlines for all points, which also wrap around on themselves.

In all cases, I am a) getting white streamlines which b) do not appear to be terminating early. However, in one case I did get a white volume plot when I was expecting red. This is happening only for the `RayCastingIntegeration` method regardless of ordering of plots. So, this *seems* similar to your case but doesn't quite duplicate your case either.

![image](https://user-images.githubusercontent.com/5720676/84834648-beffde80-afe6-11ea-86a3-c758414802d0.png)

Here is kinda sort what I as expecting

![image](https://user-images.githubusercontent.com/5720676/84835473-8cef7c00-afe8-11ea-844f-02358c22fdb3.png)

**Comment by knelli2** [comment]
Wow that was extremely thorough, thank you. You're right that is somewhat similar to the issue I was having, but not exactly. Some very valuable info though.

I don't think I've mentioned this before, but we are using .h5 files of the form
rho_b.file_0.h5
rho_b.file_1.h5
...
rho_b.file_95.h5

and loading them in as a database. While I doubt this is the issue, do you think that could be causing it? I've switched to using ribbons in my plots for now because they work and still look good, but I would still like to know why lines won't work. It's hard to get a direct comparison from the supercomputer using VisIt 3.0.0 to our local machines because for Ubuntu 16.04, the only available download was 3.0.2. Do you also think it's possible that 3.0.0 has a small bug that was fixed in 3.0.2?

**Comment by markcmiller86** [comment]
I am reasonably confident the data source has no bearing here.

In chatting with another developer, yes, we did have problems with lighting and lines in 3.0.0.

In all my testing I've been using a 3.0.1 client and a 3.0.2 engine. If I can find the time, I will try to test a 3.0.0 client/engine.

**Comment by knelli2** [comment]
Thank you Mark, you've been extremely helpful! If you can't find the time, don't worry about it. I'm pretty satisfied with there being a small bug in 3.0.0. In my opinion, it wasn't very smart for a supercomputer to use the first release of a completely new version of VisIt, and not have the subsequent patches. However, this thread has narrowed down the problem significantly and I've found a useful workaround for my plots.

**Comment by markcmiller86** [comment]
Ok, I think I have sufficient justification now to close this issue.

We no longer have 3.0.0 installed on our servers.  So, I downloaded 3.0.0 and ran it all locally on my laptop and confirmed that at least *some* of the streamlines, those that overlap with the volume plot, appear to have been colored incorrectly. But, the problem appears to be corrected in 3.0.1. See the images below.

### 3.0.0

![image](https://user-images.githubusercontent.com/5720676/84862551-3ace4b00-b028-11ea-9fa2-1d4f137dda9e.png)

### 3.0.1

![image](https://user-images.githubusercontent.com/5720676/84862506-27bb7b00-b028-11ea-90a9-528671c93a61.png)
